aws/src/main/java/org/apache/iceberg/aws/s3/S3InputStream.java

```diff
@@ -0,0 +1,168 @@
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+
+class S3InputStream extends SeekableInputStream {
+  private static final Logger LOG = LoggerFactory.getLogger(S3InputStream.class);
+
+  private final StackTraceElement[] createStack;
+  private final S3Client s3;
+  private final S3URI location;
+  private final AwsProperties awsProperties;
+
+  private InputStream stream;
+  private long pos = 0;
+  private long next = 0;
+  private boolean closed = false;
+
+  private int skipSize = 1024 * 1024;
+
+  S3InputStream(S3Client s3, S3URI location) {
+    this(s3, location, new AwsProperties());
+  }
+
+  S3InputStream(S3Client s3, S3URI location, AwsProperties awsProperties) {
+    this.s3 = s3;
+    this.location = location;
+    this.awsProperties = awsProperties;
+
+    createStack = Thread.currentThread().getStackTrace();
+  }
+
+  @Override
+  public long getPos() {
+    return next;
+  }
+
+  @Override
+  public void seek(long newPos) {
+    Preconditions.checkState(!closed, "already closed");
+    Preconditions.checkArgument(newPos >= 0, "position is negative: %s", newPos);
+
+    // this allows a seek beyond the end of the stream but the next read will fail
+    next = newPos;
+  }
+
+  @Override
+  public int read() throws IOException {
+    Preconditions.checkState(!closed, "Cannot read: already closed");
+    positionStream();
+
+    pos += 1;
+    next += 1;
+
+    return stream.read();
+  }
+
+  @Override
+  public int read(byte[] b, int off, int len) throws IOException {
+    Preconditions.checkState(!closed, "Cannot read: already closed");
+    positionStream();
+
+    int bytesRead = stream.read(b, off, len);
+    pos += bytesRead;
+    next += bytesRead;
+
+    return bytesRead;
+  }
+
+  @Override
+  public void close() throws IOException {
+    super.close();
+    closed = true;
+    closeStream();
+  }
+
+  private void positionStream() throws IOException {
+    if ((stream != null) && (next == pos)) {
+      // already at specified position
+      return;
+    }
+
+    if ((stream != null) && (next > pos)) {
+      // seeking forwards
+      long skip = next - pos;
+      if (skip <= Math.max(stream.available(), skipSize)) {
+        // already buffered or seek is small enough
+        LOG.debug("Read-through seek for {} to offset {}", location, next);
+        try {
+          ByteStreams.skipFully(stream, skip);
+          pos = next;
+          return;
+        } catch (IOException ignored) {
+          // will retry by re-opening the stream
+        }
+      }
+    }
+
+    // close the stream and open at desired position
+    LOG.debug("Seek with new stream for {} to offset {}", location, next);
+    pos = next;
+    openStream();
+  }
+
+  private void openStream() throws IOException {
+    GetObjectRequest.Builder requestBuilder = GetObjectRequest.builder()
+        .bucket(location.bucket())
+        .key(location.key())
+        .range(String.format("bytes=%s-", pos));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

This code uses &#39;%s&#39; to format int: pos. This is a potential locale-sensitive handling issue. It might cause errors in the handling and processing of the statement at line: 139. Consider formatting this data with &#39;%d&#39; instead.